### ISSUE TITLE
no-unnecessary-generics: false negatives with unrelated Identifiers or shadowing

### ISSUE THREAD

**Issue by ajafff** [Microsoft/dtslint#131]
Ref: https://github.com/Microsoft/TypeScript/issues/26129#issuecomment-410050853

`no-unnecessary-generics` does not report certain TypeParameters if an Identifier with the same name is used somewhere within a parameter's type:

```ts
// unrelated identifier
declare function test<T>(param: {T: any}): any;

// shadowing
declare function test<T>(param: {[T in 'foo']: T}): any;

declare function test<T>(param: Foo extends () => infer T ? T : never): any;
```

**Comment by anion155** [comment]
@andy-ms Still can be reproduced with 0.5.7:
```typescript
interface ReactResizeDetectorProps<
    R extends RefreshMode
> {
    refreshMode?: R;
    refreshOptions?: RefreshModeOptions[R];
}
export function withResizeDetector<
    R extends RefreshMode
>(
    props?: ReactResizeDetectorProps<R>
): void;
```
